cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -32,269 +31,145 @@ import (
 
-	// At this point we have a front proxy CA signing key.  We can use that create the front proxy client cert if
-	// it doesn't already exist.
-	// If at least one of them exists, we should try to load them
-	// In the case that only one exists, there will most likely be an error anyway
-	if pkiutil.CertOrKeyExist(pkiDir, kubeadmconstants.FrontProxyClientCertAndKeyBaseName) {
-		// Try to load apiserver-kubelet-client.crt and apiserver-kubelet-client.key from the PKI directory
-		apiCert, apiKey, err := pkiutil.TryLoadCertAndKeyFromDisk(pkiDir, kubeadmconstants.FrontProxyClientCertAndKeyBaseName)
-		if err != nil || apiCert == nil || apiKey == nil {
-			return fmt.Errorf("certificate and/or key existed but they could not be loaded properly")
-		}
+	return frontProxyCACert, frontProxyCAKey, nil
+}
 
-		fmt.Println("[certificates] Using the existing front-proxy client certificate and key.")
-	} else {
-		// The certificate and the key did NOT exist, let's generate them now
-		// TODO: Add a test case to verify that this cert has the x509.ExtKeyUsageClientAuth flag
-		config := certutil.Config{
-			CommonName: "front-proxy-client",
-			Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
-		}
-		apiClientCert, apiClientKey, err := pkiutil.NewCertAndKey(frontProxyCACert, frontProxyCAKey, config)
-		if err != nil {
-			return fmt.Errorf("failure while creating front-proxy client key and certificate [%v]", err)
-		}
+// NewFrontProxyClientCertAndKey generate CA certificate for proxy server client, signed by the given front proxy CA.
+func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProxyCAKey *rsa.PrivateKey) (*x509.Certificate, *rsa.PrivateKey, error) {
 
-		if err = pkiutil.WriteCertAndKey(pkiDir, kubeadmconstants.FrontProxyClientCertAndKeyBaseName, apiClientCert, apiClientKey); err != nil {
-			return fmt.Errorf("failure while saving front-proxy client certificate and key [%v]", err)
-		}
-		fmt.Println("[certificates] Generated front-proxy client certificate and key.")
+	config := certutil.Config{
+		CommonName: "front-proxy-client",
+		Usages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+	}
+	frontProxyClientCert, frontProxyClientKey, err := pkiutil.NewCertAndKey(frontProxyCACert, frontProxyCAKey, config)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failure while creating front-proxy client key and certificate: %v", err)
 	}
 
-	fmt.Printf("[certificates] Valid certificates and keys now exist in %q\n", pkiDir)
-
-	return nil
+	return frontProxyClientCert, frontProxyClientKey, nil
 }
 
-// checkAltNamesExist verifies that the cert is valid for all IPs and DNS names it should be valid for
-func checkAltNamesExist(IPs []net.IP, DNSNames []string, altNames certutil.AltNames) bool {
-	dnsset := setutil.NewString(DNSNames...)
+// getAltNames builds an AltNames object for to be used when generating apiserver certificate
+func getAltNames(cfg *kubeadmapi.MasterConfiguration) (*certutil.AltNames, error) {
```

Comment:
Frankly speaking this was something inherited from the original implementation.
However, if I change from pointer to struct I will be forced to return some kind of empty/structure in case of errors instead of nil, and IMO this will make the code less clear.
Let me know...